File: client/project-fill/Assets/Scripts/InGame/BoardCodec.cs
using System.Collections.Generic;
using System.Text;

namespace Game.InGame
{
    // Encodes/decodes the compact board layout stored in stage.csv `board` column.
    // Fixed Capacity chars per lane, no delimiters (read in groups of Capacity, lane order = csv order):
    //   uppercase glyph = normal chip   (R B G Y P C O M L T = SignalType 0..9)
    //   lowercase glyph = overload chip
    //   '-'             = empty slot     (empty lane = "----")
    // Lane kind / locked / unlock come from the StageDefinition (lane_kinds + lock_unlock), not from here.
    // Mirror of tools/stage_editor/src/lib/board-codec.ts — keep both in sync.
    public static class BoardCodec
    {
        private const string Glyphs = "RBGYPCOMLT";

        public static Board Decode(string board, StageDefinition def)
        {
            int laneCount = def.LaneCount;
            var lanes = new List<SlotLane>(laneCount);
            for (int i = 0; i < laneCount; i++)
            {
                var kind   = def.LaneKinds[i];
                var unlock = def.LockUnlock != null && i < def.LockUnlock.Length ? def.LockUnlock[i] : SignalType.Red;
                var lane   = new SlotLane(kind, unlock);

                int baseIdx = i * SlotLane.Capacity;
                for (int s = 0; s < SlotLane.Capacity; s++)
                {
                    int p = baseIdx + s;
                    if (board == null || p >= board.Length) break;
                    char ch = board[p];
                    if (ch == '-') continue;
                    int t = Glyphs.IndexOf(char.ToUpperInvariant(ch));
                    if (t < 0) continue;
                    lane.Push(new Chip((SignalType)t, char.IsLower(ch)));
                }
                lanes.Add(lane);
            }

            var relay = def.RelayOrder != null ? new List<SignalType>(def.RelayOrder) : null;
            return new Board(lanes, def.Types, relay);
        }

        public static string Encode(Board board)
        {
            var sb = new StringBuilder(board.Lanes.Count * SlotLane.Capacity);
            foreach (var lane in board.Lanes)
            {
                for (int s = 0; s < SlotLane.Capacity; s++)
                {
                    if (s < lane.Chips.Count)
                    {
                        var c = lane.Chips[s];
                        char g = Glyphs[(int)c.Type];
                        sb.Append(c.Overload ? char.ToLowerInvariant(g) : g);
                    }
                    else sb.Append('-');
                }
            }
            return sb.ToString();
        }
    }
}

File: tools/stage_editor/src/lib/board-codec.ts
// Compact board encoding stored in stage.csv `board` column.
// Fixed 4 chars per lane (Capacity), no delimiters: read in groups of 4, lane order = csv order.
//   uppercase glyph = normal chip   (R B G Y P C O M L T)
//   lowercase glyph = overload chip  (overload only ever applies to the overload_type)
//   '-'             = empty slot     (an empty lane = "----")
// Lane kind / locked / unlock come from the lane_kinds + lock_unlock columns, not from here.

import type { LaneData, ChipData } from '../types/stage';
import { SIGNAL_GLYPHS, LaneKind } from './signal';

const SLOTS = 4;

export function encodeBoard(lanes: LaneData[]): string {
  return lanes
    .map(l => {
      let s = '';
      for (let i = 0; i < SLOTS; i++) {
        const c = l.chips[i];
        if (!c) { s += '-'; continue; }
        const g = SIGNAL_GLYPHS[c.type] ?? '?';
        s += c.overload ? g.toLowerCase() : g;
      }
      return s;
    })
    .join('');
}

export function decodeBoard(board: string, kinds: number[], unlock: number[]): LaneData[] {
  const lanes: LaneData[] = [];
  for (let i = 0; i < kinds.length; i++) {
    const seg = board.slice(i * SLOTS, i * SLOTS + SLOTS);
    const chips: ChipData[] = [];
    for (const ch of seg) {
      if (ch === '-') continue;
      const idx = SIGNAL_GLYPHS.indexOf(ch.toUpperCase());
      if (idx < 0) continue;
      const overload = ch === ch.toLowerCase() && ch !== ch.toUpperCase();
      chips.push({ type: idx, overload });
    }
    lanes.push({
      kind: kinds[i],
      locked: kinds[i] === LaneKind.Locked,
      unlockType: unlock[i] >= 0 ? unlock[i] : 0,
      chips,
    });
  }
  return lanes;
}

export function hasBoard(board: string | undefined | null): boolean {
  return !!board && /[^-]/.test(board);
}
